### PR TITLE
Datasource: Overhaul plugin error handling and action buttons

### PR DESCRIPTION
**Pull request by sasklacz**
Fixes https://github.com/grafana/enterprise-datasources/issues/158
Fixes https://github.com/grafana/enterprise-datasources/issues/159

**Special notes for your reviewer:**

~This is a work in progress just to check if this approach is correct.~ Since we want to skip the usual flow, I've used the parameters that we already have and based on them sent new events - from the backend server - which are handled by the `EditDataSource` page. This way it's fully backwards compatible and contained to this use case ~(one thing that comes to my mind, is that those events potentially can still be triggered by user actions in other parts of the application (because of using the existing parameters for this logic) and then the page would load with the error from some other flow?)~. I've also created a new `Alert` component just for this purpose as we're changing the icon and styling in the testing status section and otherwise this could potentially influence other parts of the application. There's also a hardcoded message and links to explore/dashboards, so I don't really see much sense in trying to incorporate that into the existing component.

~Additionally when starting the data source test, we set the status to 'info', which the `Alert` component supports but we were just limiting the options to `success` and `error`. Imo it looks good, but currently there's a small glitch when the alert loads with previous status styling before switching to info which I would have to address if we want to keep it. Or I can just leave it as it is.~
Talked with @catherineymgui and she's ok with using it. @torkelo had no opinion.

Screenshot 2023-04-21 at 03 23 32

When the user has no rights to use Explore, the link button is grayed out and pointer events are disabled:

Screenshot 2023-04-23 at 22 45 44

Loom video showing how it works:
https://www.loom.com/share/bd140744cfdc477d8d8c413d4e236d1e

# Release notice breaking change

The type signature of the `testDatasource()` method on the `DataSourceWithBackend` class [has changed](https://github.com/grafana/grafana/pull/67014/files/a5608dc4f27ab4459e725b22ff60b8fc05390c08#diff-c58fc1a09e9b9b17e5f45efbfb646273e69145f7687facb134440da4edafc745R263), the returned Promise is now typed stricter, which is probably going to cause type-errors while building plugins against the latest Grafana versions.

```typescript
// Before
abstract testDatasource(): Promise<any>;

// After
abstract testDatasource(): Promise<TestDataSourceResponse>;
```

### CODE REVIEW

**Comment by catherineymgui** [comment]
- In the loom video: At 7 seconds in — the error message switches to “success” for a brief second before showing the loading / info alert. Why is this happening?
- Similarly at 15 seconds in — the error message switches to “error” for a brief second before showing the loading / info alert. Same question as ^
- How long could testing take? Can we change the copy to say something like “Performing plugin health check…this could take as long as <x> min” so users can anticipate how long this could take?

**Comment by sasklacz** [comment]
> * In the loom video: At 7 seconds in — the error message switches to “success” for a brief second before showing the loading / info alert. Why is this happening?
> 
>     * Similarly at 15 seconds in — the error message switches to “error” for a brief second before showing the loading / info alert. Same question as ^
> 

Sorry, this was a borked version and I updated that in the meantime. Can you recheck https://www.loom.com/share/bd140744cfdc477d8d8c413d4e236d1e ?

>     * How long could testing take? Can we change the copy to say something like “Performing plugin health check…this could take as long as  min” so users can anticipate how long this could take?

The longest I've seen is maybe 20s ?

**Comment by catherineymgui** [comment]
> The longest I've seen is maybe 20s ?

20s is pretty long and since we don't know how long this may take for other data sources, let's add in the text: "Testing... this could take up to a couple minutes"? 
 
Also - I don't know if we talked about this for this round of error handling, but before we discussed having inline errors be shown on validation/testing instead of on default. Can this be done (not sure if feasible for this PR)?

**Comment by sasklacz** [comment]
> 20s is pretty long and since we don't know how long this may take for other data sources, let's add in the text: "Testing... this could take up to a couple minutes"?
> 

👍 

> Also - I don't know if we talked about this for this round of error handling, but before we discussed having inline errors be shown on validation/testing instead of on default. Can this be done (not sure if feasible for this PR)?

This will come in next PRs

File: public/app/core/services/backend_srv.ts
Comment (leventebalogh) [inline public/app/core/services/backend_srv.ts:308]: Does this mean that if any call-site suppresses success alerts (`config.showSuccessAlert = false`) and the data object has a `message` property other than `"Datasource updated"`, then it is going to emit a `DataSourceConfigEvents.success` event?

File: public/app/core/services/backend_srv.ts
Comment (leventebalogh) [inline public/app/core/services/backend_srv.ts:352]: Does this mean that every time the error alert is suppressed, we are going to emit a `DataSourceConfigEvents.error` event? As this service is not only used by data-source configs this doesn't feel legit.

I think we shouldn't introduce datasource-config related knowledge in the BackendSrv, as it tries to be a generic service for interacting with the backend (even if it might has some broken windows in that regard).

Why don't we handle these alerts (success or error) at the call-site? I think we could get the same information from the observable?

File: public/app/core/services/backend_srv.ts
Comment (leventebalogh) [inline public/app/core/services/backend_srv.ts:368]: Typo
```suggestion
    // if showErrorAlert is undefined we only show alerts non data query and local api requests
```

**Review by leventebalogh — CHANGES_REQUESTED** [review]

File: public/app/core/services/backend_srv.ts
Comment (torkelo) [reply public/app/core/services/backend_srv.ts:308]: the biggest question is why is there a DataSourceConfigEvents.success event, and why is emitted from BackendSrv?

File: public/app/features/datasources/api.ts
Comment (torkelo) [inline public/app/features/datasources/api.ts:78]: I am not seeing a try-catch where updateDataSource is called that handles the error

File: public/app/features/datasources/components/DataSourceConfigAlert.tsx
Comment (torkelo) [inline public/app/features/datasources/components/DataSourceConfigAlert.tsx:24]: Why duplicate this whole component?

File: public/app/features/datasources/components/EditDataSource.tsx
Comment (torkelo) [reply public/app/features/datasources/components/EditDataSource.tsx:70]: appEvents is a legacy feature that we should not build new features around. 

If you want to capture errors from backendSrv calls use try/catch, or if you use backendSrv.fetch (that returns observable) use the observable operator catchError

**Review by torkelo — CHANGES_REQUESTED** [review]
* I don't think this change to data source settings page needs any changes to backendSrv
* Does not need a full duplicate of the Alert component 
* Does not need new appEvents 

What it could need
* Add some proper typing to `DataSourceApi.testDatasource: Promise<any>` (a Promise that returns an object with a status: error | string, and message) 
* If we don't want to saving a data source config to show an app alert we need to set showSuccessAlert false and handle errors using try/catch or cachError, and update redux state . 
* Unify the  data source testing state and saving/save success state and use the same Alert for that

File: public/app/features/datasources/components/EditDataSource.tsx
Comment (sasklacz) [reply public/app/features/datasources/components/EditDataSource.tsx:70]: Yeah, I didn't know it's a legacy thing. Just wanted something that could work without disturbing the rest of the app.

File: packages/grafana-data/src/types/datasource.ts
Comment (sasklacz) [reply packages/grafana-data/src/types/datasource.ts:23]: Not gonna use rxjs events so this will go away.

File: public/app/core/services/backend_srv.ts
Comment (sasklacz) [reply public/app/core/services/backend_srv.ts:308]: I just copied what we do for other errors/success by emmiting an event, but since it's a legacy thing I'll just do it with try/catch.

File: public/app/features/datasources/components/DataSourceConfigAlert.tsx
Comment (sasklacz) [reply public/app/features/datasources/components/DataSourceConfigAlert.tsx:24]: My idea was to not add more complexity to the existing one, but I can just merge them

File: public/app/features/datasources/api.ts
Comment (sasklacz) [reply public/app/features/datasources/api.ts:78]: @torkelo regarding

>Add some proper typing to DataSourceApi.testDatasource: Promise<any> (a Promise that returns an object with a status: error | string, and message)

you mean adding a return type here massaging the data that comes from the `put` call, right ?

**Comment by sasklacz** [comment]
@torkelo @leventebalogh Can you re-check now while I'm fixing the tests ? One thing I'm wondering about is if I should also change the return type for `testDatasource` for all of the core plugins ?

File: packages/grafana-ui/src/components/Alert/Alert.tsx
Comment (torkelo) [inline packages/grafana-ui/src/components/Alert/Alert.tsx:27]: please remove, we do not want to redesign this component with this change, must be the figma design that is mistaken

File: public/app/features/datasources/components/EditDataSource.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDataSource.tsx:126]: ?

**Review by torkelo — COMMENTED** [review]
still some confusing changes :) 

* Still changes in backend_srv 
* Unnecessary one-off design changes to a very central design system component (Alert)   
* Missing any feedback that the data source have been saved 
* Missing app notification when you delete a data source

File: public/app/features/datasources/state/reducers.ts
Comment (nmarrs) [inline public/app/features/datasources/state/reducers.ts:149]:
```suggestion
        message: 'Testing... this could take a couple of minutes',
```

**Review by nmarrs — COMMENTED** [review]
Great job on this @sasklacz! Only have a few nits to add to other reviewers' feedback 👍

File: packages/grafana-data/src/types/datasource.ts
Comment (leventebalogh) [inline packages/grafana-data/src/types/datasource.ts:479]: Nice that you are tidying that `any` up! 👏 

I think it would be semantically more correct to keep the message being just a "string", and have an optional property for the error itself. (Then we would get rid of the comment as well.)

```typescript
export interface TestDataSourceResponse {
    status: string;
    message: string;
    error?: Error;
}
```

File: public/app/core/services/backend_srv.ts
Comment (leventebalogh) [inline public/app/core/services/backend_srv.ts:335]: Why did we bring this "bail-out-early" down here? (It happened first, because if the condition is true then we don't need assign any variables or evaluate anything else.)

File: public/app/core/services/backend_srv.ts
Comment (leventebalogh) [inline public/app/core/services/backend_srv.ts:349]: I would do any changes like these in a separate, more atomic PR (it's easier to reason about them separately and probably more stakeholders will be involved).

Although a change like this (changing the error message and moving the info to the description) can make sense, it can affect different parts of the app or even plugins - the error emitted through the event bus, and if a certain part of the app was only using the error message until now, then after this change it would lose information on the UI.

File: public/app/features/datasources/components/ButtonRow.tsx
Comment (leventebalogh) [inline public/app/features/datasources/components/ButtonRow.tsx:22]: Just pinging @teddyba and @mikkancso here, since the "Explore" button was added not long ago, is the decision to remove it agreed on between the teams?

File: public/app/features/datasources/state/actions.ts
Comment (leventebalogh) [inline public/app/features/datasources/state/actions.ts:157]: What's the reason to pass in a copy of `formattedError` here?

File: public/app/features/datasources/state/actions.ts
Comment (leventebalogh) [inline public/app/features/datasources/state/actions.ts:240]: What actually does this todo mean?

File: public/app/features/datasources/state/actions.ts
Comment (leventebalogh) [inline public/app/features/datasources/state/actions.ts:275]: I guess `type` can only be a `string`? Probably we could also use `unknown` instead of `any` for `payload` as well.

File: public/app/features/datasources/components/ButtonRow.tsx
Comment (mikkancso) [reply public/app/features/datasources/components/ButtonRow.tsx:22]: Thanks for pinging us, @leventebalogh.
I'm not aware of any common decision. What is the rationale behind removing this?

File: public/app/core/services/backend_srv.ts
Comment (sasklacz) [reply public/app/core/services/backend_srv.ts:335]: Yeah sorry, I reverted my commits but guess something got borked in the process. I'll just bring this back to what it was before as we're not using anything here anymore

File: packages/grafana-ui/src/components/Alert/Alert.tsx
Comment (sasklacz) [reply packages/grafana-ui/src/components/Alert/Alert.tsx:76]: I haven't used some of the core plugins too much so I haven't seen all the cases of alerts but I can just put it in the same div and in case something is off we'll fix it while testing.

File: packages/grafana-ui/src/components/Alert/Alert.tsx
Comment (sasklacz) [reply packages/grafana-ui/src/components/Alert/Alert.tsx:153]: Yeah we need to grab the designers after 10 is out and figure things like that out. Looks like there's some missed opportunity for a better communication there.

File: public/app/features/datasources/components/ButtonRow.tsx
Comment (sasklacz) [reply public/app/features/datasources/components/ButtonRow.tsx:22]: This has been discussed month ago for the config redesign project. We replace the button with a link in the alert, and there's also a button in the header. cc @catherineymgui 

Screenshot 2023-04-23 at 22 45 44

File: public/app/features/datasources/components/DataSourceTestingStatus.tsx
Comment (sasklacz) [reply public/app/features/datasources/components/DataSourceTestingStatus.tsx:66]: Yeah I just used what was recently added to the header but makes sense

File: public/app/features/datasources/state/actions.ts
Comment (sasklacz) [reply public/app/features/datasources/state/actions.ts:157]: Sorry changed in the call above, but missed here. Thanks for catching it

File: public/app/features/datasources/components/ButtonRow.tsx
Comment (catherineymgui) [reply public/app/features/datasources/components/ButtonRow.tsx:22]: Yup, the rationale: You can only explore and create a dashboard for this data source once you have a successful health check. Thus having it as an action you can take on the same hierarchy of save and test is confusing (since you can only explore after saving and test). So the idea is that once this health check is successful, someone can perform this action. Like Kuba said, these are now showing as buttons in the header and upon a succesful health check

File: public/app/features/datasources/state/actions.ts
Comment (sasklacz) [reply public/app/features/datasources/state/actions.ts:240]: I saw this kind of comments in other places. Here the name is `namdExits` instead of `nameExists` but I don't want to fix it in the same PR.

File: public/app/features/datasources/state/actions.ts
Comment (sasklacz) [reply public/app/features/datasources/state/actions.ts:283]: Do we want a separate action or just rename this one to cover both cases?

**Comment by sasklacz** [comment]
> still some confusing changes :)
> 
>     * Still changes in backend_srv
> 
>     * Unnecessary one-off design changes to a very central design system component (Alert)
> 
>     * Missing any feedback that the data source have been saved
> 
>     * Missing app notification when you delete a data source

Do we need a notification that data source was saved if we already have the 'Testing...' status shown every time user clicks the button anyway ?

**Comment by torkelo** [comment]
> Do we need a notification that data source was saved if we already have the 'Testing...' status shown every time user clicks the button anyway ?

We usually have a toast or some feedback yes. But not sure if it's required here. Maybe if that Alert box said something like 

* Saving.... 

Changes to 

* Saved
* Testing....

Changes to 

* Saved
* Test success
* Resut of test success message...  

Maybe strange to have in case of error

**Comment by catherineymgui** [comment]
@torkelo I do like your idea but would the save action / completion will be almost instant and thus easy for an user to miss? @sasklacz can you double check how long someone might be able to see a save message for?

But the idea is that since the button itself is encompassing saving and testing, the resulting success / error message encompasses saving and testing. The assumption is that the result of the test is ultimately more important than if it was saved (thus if the error was on the test, but save succeeded, then it's an 'error' mentioning it was saved.)

**Comment by sasklacz** [comment]
> @torkelo I do like your idea but would the save action / completion will be almost instant and thus easy for an user to miss? @sasklacz can you double check how long someone might be able to see a save message for?
> 
> But the idea is that since the button itself is encompassing saving and testing, the resulting success / error message encompasses saving and testing. The assumption is that the result of the test is ultimately more important than if it was saved (thus if the error was on the test, but save succeeded, then it's an 'error' mentioning it was saved.)

Yes, the save is pretty much instantaneous so this would cause the alert to flicker when quickly changing from green save to blue info.

File: public/app/features/datasources/components/ButtonRow.tsx
Comment (mikkancso) [reply public/app/features/datasources/components/ButtonRow.tsx:22]: Thanks for the explanation! Good stuff!

**Comment by torkelo** [comment]
@sasklacz don't force push if you don't have to on larger PRs :)  makes reviewing / testing PRs a bit painful as can't pull the latest without merge conflict and have to git reset

File: packages/grafana-data/src/types/datasource.ts
Comment (torkelo) [inline packages/grafana-data/src/types/datasource.ts:481]: @leventebalogh this could lead to new typescript errors in data sources, is that ok in a major version?

**Review by torkelo — APPROVED** [review]
Pushed a minor update, that fixes some styling & whitespace issues in the Alert success text. 

Also updated the wording a bit to read better. @catherineymgui does it look ok? 

Screenshot from 2023-04-28 08-56-19

File: packages/grafana-data/src/types/datasource.ts
Comment (leventebalogh) [reply packages/grafana-data/src/types/datasource.ts:481]: Good catch. Yeah, I think as it is "only" going to cause a compile-time error, it should break existing plugins, and only needs to be addressed when someone is updating a plugin - which should be fine in a major version I think. What we should probably do is marking it as a breaking change (add it to the CHANGELOG) and mention it in our migration guide.

**Review by leventebalogh — APPROVED** [review]
LGTM, I only have one UX concern, when I scroll down and click on the "Save & test", I won't see the notification (alert) below it - it is rendered "outside of the viewport". I think this is something that can cause confusion now that we removed the default error / success notifications.

Maybe we could add some spacing to the bottom of the form, so they can "scroll further down" before clicking on save and thus see the alert rendered afterwards?  

https://user-images.githubusercontent.com/9974811/235085640-67e1e73f-8063-4e63-8606-7f4d3d3cc77a.mov

**Comment by torkelo** [comment]
hm.. that looks like a big ux issue if you have to scroll to see the save feedback (test result). Quick fix would be to move it back to where it was above the button

File: packages/grafana-data/src/types/datasource.ts
Comment (leventebalogh) [reply packages/grafana-data/src/types/datasource.ts:481]: **FYI:** I have added the "breaking change" label and updated the PR description with a breaking change notice message.

**Comment by sasklacz** [comment]
> @sasklacz don't force push if you don't have to on larger PRs :) makes reviewing / testing PRs a bit painful as can't pull the latest without merge conflict and have to git reset

Yeah sorry for that. At once point merged with main via gh UI and then couldn't really untangle it in other way

> hm.. that looks like a big ux issue if you have to scroll to see the save feedback (test result). Quick fix would be to move it back to where it was above the button

@catherineymgui thoughts ?

File: packages/grafana-data/src/types/datasource.ts
Comment (sasklacz) [reply packages/grafana-data/src/types/datasource.ts:481]: Thanks for the edits. I think this is a minor breaking change but cleans up the flow a bit in return.

**Comment by sasklacz** [comment]
@leventebalogh How do I fix the breaking changes test and will be release be from the main and I can add the `no-backport` label here ?

**Comment by catherineymgui** [comment]
> Pushed a minor update, that fixes some styling & whitespace issues in the Alert success text.
> 
> Also updated the wording a bit to read better. @catherineymgui does it look ok?
> 
> Screenshot from 2023-04-28 08-56-19

@torkelo this updated copy is great. Thank you!

**Pull request merged**
